### PR TITLE
[FLINK-33210] Introduce lineage graph relevant interfaces

### PR DESCRIPTION
**Pull request by FangYongs**
## What is the purpose of the change

This PR aims to add lineage graph relevant interfaces such as `LineageGraph`, `LineageVertex`, `SourceLineageVertex` and `LineageEdge`

## Brief change log
  - Add lineage graph relevant interfaces
  - Add default lineage graph implementation

## Verifying this change

This change added tests and can be verified as follows:

  - Added test `LineageGraphTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / no) no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / no) yes
  - The serializers: (yes / no / don't know) no
  - The runtime per-record code paths (performance sensitive): (yes / no / don't know) no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / no / don't know) no
  - The S3 file system connector: (yes / no / don't know) no

## Documentation

  - Does this pull request introduce a new feature? (yes / no) yes
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented) not applicable

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/lineage/DefaultLineageGraph.java [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+    @Override
+    public List<LineageVertex> sinks() {
+        return sinks;
+    }
+
+    @Override
+    public List<LineageEdge> relations() {
+        return lineageEdges;
+    }
```
Comment (KarmaGYZ) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/lineage/DefaultLineageGraph.java:59]: If these should not be changed outside, we may use `Collections.unmodifiableList` to guard them.

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/lineage/LineageGraphTest.java [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/** Testing for lineage graph. */
+class LineageGraphTest {
```
Comment (KarmaGYZ) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/lineage/LineageGraphTest.java:32]:
```suggestion
class DefaultLineageGraphTest {
```

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/lineage/LineageGraphTest.java [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        LineageVertex sink2 = new TestingLineageVertex("sink2");
+        LineageGraph lineageGraph =
+                DefaultLineageGraph.builder()
+                        .addLineageEdge(new TestingLineageEdge(source1, sink1))
+                        .addLineageEdges(
+                                new TestingLineageEdge(source2, sink2),
+                                new TestingLineageEdge(source3, sink1),
+                                new TestingLineageEdge(source1, sink2))
+                        .build();
+        assertThat(lineageGraph.sources()).isEqualTo(Arrays.asList(source1, source2, source3));
```
Comment (KarmaGYZ) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/lineage/LineageGraphTest.java:48]:
```suggestion
        assertThat(lineageGraph.sources()).containsExactlyInAnyOrder(source1, source2, source3);
```

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/lineage/LineageGraphTest.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        assertThat(lineageGraph.sinks()).isEqualTo(Arrays.asList(sink1, sink2));
```
Comment (KarmaGYZ) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/lineage/LineageGraphTest.java:49]:
```suggestion
        assertThat(lineageGraph.sinks()).containsExactlyInAnyOrder(sink1, sink2);
```

**Review by KarmaGYZ — COMMENTED** [review]
Thanks for the PR. Will merge it after all the comments addressed.

**Review by KarmaGYZ — APPROVED** [review]

**Comment by FangYongs** [comment]
@flinkbot run azure

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/lineage/DefaultLineageGraph.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+public class DefaultLineageGraph implements LineageGraph {
+    private final List<LineageEdge> lineageEdges;
+    private final List<SourceLineageVertex> sources;
+    private final List<LineageVertex> sinks;
+
+    private DefaultLineageGraph(List<LineageEdge> lineageEdges) {
+        this.lineageEdges = lineageEdges;
+
+        Set<SourceLineageVertex> srcs = new HashSet<>();
+        Set<LineageVertex> targets = new HashSet<>();
```
Comment (JingGe) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/lineage/DefaultLineageGraph.java:41]: NIT: 
```suggestion
        Set<LineageVertex> deduplicatedSinks = new HashSet<>();
```

**Review by JingGe — COMMENTED** [review]
Thanks for driving it! I just left some NIT comments

**Comment by FangYongs** [comment]
Thanks @JingGe , please help to review this PR again

**Comment by FangYongs** [comment]
@HuangZhenQiu @mobuchowski I have add dataset and facet interfaces in lineage vertex, please help to review this PR cc @X-czh

File: flink-streaming-java/src/test/java/org/apache/flink/streaming/api/lineage/DefaultLineageGraphTest.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+                DefaultLineageGraph.builder()
+                        .addLineageEdge(new TestingLineageEdge(source1, sink1))
+                        .addLineageEdges(
+                                new TestingLineageEdge(source2, sink2),
+                                new TestingLineageEdge(source3, sink1),
+                                new TestingLineageEdge(source1, sink2))
+                        .build();
+        assertThat(lineageGraph.sources()).containsExactlyInAnyOrder(source1, source2, source3);
+        assertThat(lineageGraph.sinks()).containsExactlyInAnyOrder(sink1, sink2);
+        assertThat(lineageGraph.relations().size()).isEqualTo(4);
```
Comment (X-czh) [inline flink-streaming-java/src/test/java/org/apache/flink/streaming/api/lineage/DefaultLineageGraphTest.java:52]: NIT: you can use `assertThat(lineageGraph.relations()).hasSize(4)` here

**Comment by X-czh** [comment]
@FangYongs Thanks for the PR. The changes LGTM except for a nit on the UT.

**Review by mobuchowski — APPROVED** [review]
Looks good to me.

**Pull request merged**